modules/topicsFpdModule.js

```diff
@@ -73,6 +114,145 @@ export function processFpd(config, {global}, {data = topicsData} = {}) {
+ */
+function receiveMessage(evt) {
+  if (evt && evt.data) {
+    try {
+      let data = safeJSONParse(evt.data);
+      if (includes(getLoadedIframeURL(), evt.origin) && data && data.segment && !isEmpty(data.segment.topics)) {
+        const {domain, topics, bidder} = data.segment;
+        const iframeTopicsData = getTopicsData(domain, topics)[0];
+        iframeTopicsData && storeInLocalStorage(bidder, iframeTopicsData);
+      }
+    } catch (err) { }
+  }
+}
+
+/**
+Function to store Topics data recieved from iframe in storage(name: "prebid:topics")
+* @param {Topics} topics
+*/
+export function storeInLocalStorage(bidder, topics) {
+  const storedSegments = new Map(safeJSONParse(coreStorage.getDataFromLocalStorage(topicStorageName)));
+  if (storedSegments.has(bidder)) {
+    storedSegments.get(bidder)[topics['ext']['segclass']] = topics;
+    storedSegments.get(bidder)[lastUpdated] = new Date().getTime();
+    storedSegments.set(bidder, storedSegments.get(bidder));
+  } else {
+    storedSegments.set(bidder, {[topics.ext.segclass]: topics, [lastUpdated]: new Date().getTime()})
+  }
+  coreStorage.setDataInLocalStorage(topicStorageName, JSON.stringify([...storedSegments]));
+}
+
+function isCachedDataExpired(storedTime, cacheTime) {
+  const _MS_PER_DAY = 1000 * 60 * 60 * 24;
+  const now = new Date().getTime();
+  const daysDifference = Math.floor((storedTime - now) / _MS_PER_DAY);
+  return daysDifference > cacheTime;
+}
+
+/**
+* Function to get random bidders based on count passed with array of bidders
+**/
+function getRandomBidders(arr, count) {
+  return ([...arr].sort(() => 0.5 - Math.random())).slice(0, count)
+}
+
+/**
+ * function to add listener for message receiving from IFRAME
+ */
+function listenMessagesFromTopicIframe() {
+  window.addEventListener('message', receiveMessage, false);
+}
+
+function checkTCFv2(vendorData, requiredPurposes = TCF_REQUIRED_PURPOSES) {
+  const {gdprApplies, purpose} = vendorData;
+  if (!gdprApplies) {
+    return true;
+  }
+  return requiredPurposes.map((purposeNo) => {
+    const purposeConsent = purpose.consents ? purpose.consents[purposeNo] : false;
+    if (purposeConsent) {
+      return true;
+    }
+    return false;
+  }).reduce((a, b) => a && b, true);
+}
+
+function hasGDPRConsent() {
+  // Check for GDPR consent for purpose 1,2,3,4 and return false if consent has not been given
+  const gdprConsent = gdprDataHandler.getConsentData();
+  const hasGdpr = (gdprConsent && typeof gdprConsent.gdprApplies === 'boolean' && gdprConsent.gdprApplies) ? 1 : 0;
+  const gdprConsentString = hasGdpr ? gdprConsent.consentString : '';
+  if (hasGdpr) {
+    if ((!gdprConsentString || gdprConsentString === '') || !gdprConsent.vendorData) {
+      return false;
+    }
+    return checkTCFv2(gdprConsent.vendorData);
+  }
+  return true;
+}
+
+/**
+ * function to load the iframes of the bidder to load the topics data
+ */
+function loadTopicsForBidders() {
+  HAS_GDPR_CONSENT = hasGDPRConsent();
```

Comment:
@patmmccann  Can you please review the consent logic. 
I have handled specifically Purpose [1,2,3,4] consent. If one of the consent flag is false, we don't load Iframe for any of the bidders.

We have added **topicsFpd** of 'core' module type which is neither Vendor or user id module. Same implementation as we have for **enrichmentFpd** .
```
const MODULE_NAME = 'topicsFpd';
export const coreStorage = getCoreStorageManager(MODULE_NAME);
```

Reply:
Thanks!!

Reply:
Does this also check CCPA? (need to make sure not to call if the user is opted out of "sale")

Reply:
Nothing in prebid checks that; publishers are currently responsible for ccpa enforcement

Reply:
@patmmccann Device access check added. 
My understanding is we might dont need to add CCPAconsent check if deviceAccess flag is set for true/false.

Reply:
Thanks!